TulparAudio/internal/source/SourceCollection.cpp

```diff
@@ -0,0 +1,1097 @@
+{
+    LOG_AUDIO->Debug("Source #{}: set looping {}", source, flag);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcei(static_cast<ALuint>(source), AL_LOOPING, (flag ? AL_TRUE : AL_FALSE));
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set looping: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+float SourceCollection::GetSourcePitch(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALfloat result = 0.0f;
+
+    alGetSourcef(static_cast<ALuint>(source), AL_PITCH, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = -1.0f;
+        LOG_AUDIO->Warning("Source #{}: get pitch: {:#x}", source, alErr);
+    }
+
+    return result;
+}
+
+bool SourceCollection::SetSourcePitch(SourceHandle source, float value)
+{
+    LOG_AUDIO->Debug("Source #{}: set pitch {}", source, value);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcef(static_cast<ALuint>(source), AL_PITCH, value);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set pitch: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+float SourceCollection::GetSourceGain(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALfloat result = 0.0f;
+
+    alGetSourcef(static_cast<ALuint>(source), AL_GAIN, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = -1.0f;
+        LOG_AUDIO->Warning("Source #{}: get gain: {:#x}", source, alErr);
+    }
+
+    return result;
+}
+
+bool SourceCollection::SetSourceGain(SourceHandle source, float value)
+{
+    LOG_AUDIO->Debug("Source #{}: set gain {}", source, value);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcef(static_cast<ALuint>(source), AL_GAIN, value);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set gain: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+std::array<float, 3> SourceCollection::GetSourcePosition(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALfloat x;
+    ALfloat y;
+    ALfloat z;
+
+    alGetSource3f(static_cast<ALuint>(source), AL_POSITION, &x, &y, &z);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: get position: {:#x}", source, alErr);
+
+        x = std::numeric_limits<float>::quiet_NaN();
+        y = std::numeric_limits<float>::quiet_NaN();
+        z = std::numeric_limits<float>::quiet_NaN();
+    }
+
+    return {{ static_cast<float>(x), static_cast<float>(y), static_cast<float>(z) }};
+}
+
+bool SourceCollection::SetSourcePosition(SourceHandle source, std::array<float, 3> const& vec)
```

Comment:
Shouldn't it be const and static?

Reply:
I don't see it ever changing any members or calling member functions for that matter.

Reply:
> static
I would not like this method to be called without existing collection that would inherently validate the handle.

I will add `assert` statements to make sure a collection knows about given handle. It's part of the contract for a collection to work with handles it creates.

> const
I would prefer to extend `constness` beyond what c++ provides. In this case the state of the object pointed by given `handle` will be changed.